File: parser/parsers/bbva.py
import re
import logging
from datetime import date, datetime
from typing import Optional

from parsers.base import BaseStatementParser
from models.schema import ParsedStatement, ParsedHeader, ParsedBalanceSummary, ParsedTransaction
from utils.amount_parser import parse_ars_amount, parse_percentage

logger = logging.getLogger(__name__)


def _parse_date_arg(raw: str) -> Optional[date]:
    """Parse DD/MM/YY or DD-MM-YY or DD/MM/YYYY."""
    if not raw:
        return None
    raw = raw.strip()
    for fmt in ("%d/%m/%Y", "%d-%m-%Y", "%d/%m/%y", "%d-%m-%y"):
        try:
            return datetime.strptime(raw, fmt).date()
        except ValueError:
            continue
    return None


class BBVAParser(BaseStatementParser):
    """
    Parser for BBVA Argentina credit card statements (Visa Gold).
    The PDF has a clear text layout with labeled rows for the balance summary
    and a table for transactions.
    """

    def parse(self) -> ParsedStatement:
        text = self.full_text()
        lines = [l.strip() for l in text.splitlines() if l.strip()]

        header = self._parse_header(lines)
        balance = self._parse_balance(lines)
        transactions = self._parse_transactions(lines)

        return ParsedStatement(
            header=header,
            balance_summary=balance,
            transactions=transactions,
        )

    def _parse_header(self, lines: list[str]) -> ParsedHeader:
        holder_name = ""
        account_number = None
        card_last_four = "0000"
        period_start: Optional[date] = None
        period_end: Optional[date] = None
        due_date: Optional[date] = None

        full_text = "\n".join(lines)

        # Holder name — look for "Titular:" or "SR/SRA" pattern
        for i, line in enumerate(lines):
            if re.search(r"\bTitular\b", line, re.IGNORECASE):
                # Name might be on same line or next
                m = re.search(r"Titular[:\s]+([A-ZÁÉÍÓÚÑ][A-Za-záéíóúñ\s]+)", line)
                if m:
                    holder_name = m.group(1).strip()
                elif i + 1 < len(lines):
                    holder_name = lines[i + 1].strip()
                break

        # Account number
        m = re.search(r"N[úu]mero de cuenta[:\s]+([\d\s\-]+)", full_text, re.IGNORECASE)
        if m:
            account_number = m.group(1).strip().replace(" ", "")

        # Card last four — look for **** **** **** XXXX pattern
        m = re.search(r"\*+\s*\*+\s*\*+\s*(\d{4})", full_text)
        if m:
            card_last_four = m.group(1)

        # Dates — look for "Cierre" / "Vencimiento" patterns
        # BBVA format: "26/02/26" or "26-Feb-26"
        date_pattern = r"(\d{2}[/\-]\d{2}[/\-]\d{2,4})"

        # Period end (cierre actual)
        m = re.search(r"[Cc]ierre\s+[Aa]ctual[:\s]+" + date_pattern, full_text)
        if m:
            period_end = _parse_date_arg(m.group(1))

        # Period start (cierre anterior)
        m = re.search(r"[Cc]ierre\s+[Aa]nterior[:\s]+" + date_pattern, full_text)
        if m:
            period_start = _parse_date_arg(m.group(1))

        # Due date (vencimiento)
        m = re.search(r"[Vv]encimiento[:\s]+" + date_pattern, full_text)
        if m:
            due_date = _parse_date_arg(m.group(1))

        # Fallback: find any 3 dates in sequence
        if not period_end or not due_date:
            all_dates = re.findall(date_pattern, full_text)
            parsed_dates = [_parse_date_arg(d) for d in all_dates if _parse_date_arg(d)]
            if len(parsed_dates) >= 2 and not period_end:
                period_end = parsed_dates[0]
            if len(parsed_dates) >= 3 and not due_date:
                due_date = parsed_dates[2]

        today = date.today()
        return ParsedHeader(
            bank_name="BBVA",
            holder_name=holder_name or "Titular BBVA",
            account_number=account_number,
            card_last_four=card_last_four,
            card_network="Visa",
            period_start=period_start or today,
            period_end=period_end or today,
            due_date=due_date or today,
        )

    def _parse_balance(self, lines: list[str]) -> ParsedBalanceSummary:
        full_text = "\n".join(lines)
        bs = ParsedBalanceSummary()

        def find_amount(pattern: str) -> float:
            m = re.search(pattern, full_text, re.IGNORECASE | re.MULTILINE)
            if m:
                return parse_ars_amount(m.group(1))
            return 0.0

        # Argentine amount pattern: digits with dots and comma
        amt = r"([\$\s]*[\d\.]+,\d{2})"

        bs.previous_balance = find_amount(r"[Ss]aldo\s+[Aa]nterior[^$\d]*" + amt)
        bs.payments_applied = find_amount(r"[Pp]agos?\s+[Rr]ecibidos?[^$\d]*" + amt)
        bs.total_consumption = find_amount(r"[Cc]onsumos?[^$\d]*" + amt)
        bs.commission_cuenta_full = find_amount(r"[Cc]omisi[oó]n\s+[Cc]uenta\s+[Ff]ull[^$\d]*" + amt)
        bs.sello_tax = find_amount(r"[Ss]ello[^$\d]*" + amt)
        bs.financing_interest = find_amount(r"[Ii]ntereses?\s+[Ff]inanciamiento[^$\d]*" + amt)
        bs.current_balance = find_amount(r"[Ss]aldo\s+[Aa]ctual[^$\d]*" + amt)
        bs.minimum_payment = find_amount(r"[Pp]ago\s+[Mm][íi]nimo[^$\d]*" + amt)

        # IVA / IIBB combined
        bs.iva_tax = find_amount(r"IVA[^$\d]*" + amt)
        bs.iibb_tax = find_amount(r"IIBB[^$\d]*" + amt)

        # Interest rates
        m = re.search(r"TNA[:\s]+([\d,\.]+)\s*%", full_text, re.IGNORECASE)
        if m:
            bs.tna_ars = parse_percentage(m.group(1))
        m = re.search(r"TEM[:\s]+([\d,\.]+)\s*%", full_text, re.IGNORECASE)
        if m:
            bs.tem_ars = parse_percentage(m.group(1))
        m = re.search(r"TEA[:\s]+([\d,\.]+)\s*%", full_text, re.IGNORECASE)
        if m:
            bs.tea_ars = parse_percentage(m.group(1))

        return bs

    def _parse_transactions(self, lines: list[str]) -> list[ParsedTransaction]:
        """
        BBVA transaction format (each line):
        DD/MM/YY  MERCHANT NAME    [cuota X/Y]  VOUCHER  AMOUNT_ARS  [AMOUNT_USD]
        """
        transactions = []

        # Date pattern to identify transaction rows
        date_re = re.compile(r"^(\d{2}[/\-]\d{2}[/\-]\d{2,4})\s+(.+)")
        # Amount at end of line
        amount_re = re.compile(r"([\d\.]+,\d{2})\s*$")
        # Installment pattern e.g. "3/12" or "CTA 3/12"
        installment_re = re.compile(r"\bCTA\s+(\d+)/(\d+)\b|\b(\d+)/(\d+)\b")

        for line in lines:
            m = date_re.match(line)
            if not m:
                continue

            tx_date = _parse_date_arg(m.group(1))
            if not tx_date:
                continue

            rest = m.group(2).strip()

            # Skip header/footer rows
            if any(skip in rest.upper() for skip in ["FECHA", "DESCRIPCION", "IMPORTE", "TOTAL", "SALDO"]):
                continue

            # Extract amount
            am = amount_re.search(rest)
            if not am:
                continue
            amount_ars = parse_ars_amount(am.group(1))
            rest = rest[: am.start()].strip()

            # Extract installment
            inst_current = inst_total = None
            im = installment_re.search(rest)
            if im:
                if im.group(1):
                    inst_current, inst_total = int(im.group(1)), int(im.group(2))
                else:
                    inst_current, inst_total = int(im.group(3)), int(im.group(4))
                rest = rest[: im.start()].strip()

            # What remains is merchant name + possibly voucher (numeric at end)
            voucher = None
            voucher_m = re.search(r"\s(\d{6,})$", rest)
            if voucher_m:
                voucher = voucher_m.group(1)
                rest = rest[: voucher_m.start()].strip()

            merchant = rest.strip()
            if not merchant:
                continue

            transactions.append(
                ParsedTransaction(
                    date=tx_date,
                    merchant_name=merchant,
                    voucher_number=voucher,
                    installment_current=inst_current,
                    installment_total=inst_total,
                    amount_ars=amount_ars,
                )
            )

        return transactions

File: web/src/lib/pdf-parser/bbva.ts
import { parseARS, parsePct } from "./amounts";
import { parseDate } from "./dates";
import type { ParsedStatement, ParsedHeader, ParsedBalanceSummary, ParsedTransaction } from "./types";

export function parseBBVA(text: string, lines: string[]): ParsedStatement {
  return {
    header: parseHeader(text, lines),
    balance_summary: parseBalance(lines),
    transactions: parseTransactions(lines),
    parser_version: "js-1.0.0",
  };
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

const AMT_RE = /(\d{1,3}(?:\.\d{3})*,\d{2})/;

/**
 * Find the amount for a label in the lines array.
 * Strips parenthesized numbers (base amounts), then checks same line, then next line.
 * nextLineOnly=true forces looking at the next line (used when same-line value is the base, not the tax).
 */
function findAmt(lines: string[], labelRe: RegExp, nextLineOnly = false): number {
  for (let i = 0; i < lines.length; i++) {
    if (!labelRe.test(lines[i])) continue;
    if (!nextLineOnly) {
      const stripped = lines[i].replace(/\([^)]+\)/g, "");
      const m = stripped.match(AMT_RE);
      if (m) return parseARS(m[1]);
    }
    if (i + 1 < lines.length) {
      const m = lines[i + 1].match(/^(-?)(\d{1,3}(?:\.\d{3})*,\d{2})/);
      if (m) return (m[1] === "-" ? -1 : 1) * parseARS(m[2]);
    }
  }
  return 0;
}

/** Sum all matching label occurrences (for taxes that appear multiple times, e.g. sello × 2) */
function sumAmt(lines: string[], labelRe: RegExp, nextLineOnly = false): number {
  let total = 0;
  for (let i = 0; i < lines.length; i++) {
    if (!labelRe.test(lines[i])) continue;
    if (!nextLineOnly) {
      const stripped = lines[i].replace(/\([^)]+\)/g, "");
      const m = stripped.match(AMT_RE);
      if (m) { total += parseARS(m[1]); continue; }
    }
    if (i + 1 < lines.length) {
      const m = lines[i + 1].match(/^(-?)(\d{1,3}(?:\.\d{3})*,\d{2})/);
      if (m) total += (m[1] === "-" ? -1 : 1) * parseARS(m[2]);
    }
  }
  return total;
}

/** Find a percentage rate that is on the line AFTER the label, e.g. "TNA $\n80,830 %" */
function findPct(lines: string[], labelRe: RegExp): number | undefined {
  for (let i = 0; i < lines.length; i++) {
    if (!labelRe.test(lines[i])) continue;
    if (i + 1 < lines.length) {
      const m = lines[i + 1].match(/^([\d,.]+)\s*%/);
      if (m) return parsePct(m[1]);
    }
  }
  return undefined;
}

// ─── Header ───────────────────────────────────────────────────────────────────

function parseHeader(text: string, lines: string[]): ParsedHeader {
  // Account number: "Visa Gold cuenta 1078064534"
  const accM = text.match(/cuenta\s+([\d]{6,})/i);
  const accountNumber = accM?.[1];
  // Card last four = last 4 digits of account number
  const cardLastFour = accountNumber?.slice(-4) ?? "0000";

  // Holder name: first all-caps two-word line near top that isn't a label
  let holderName = "Titular BBVA";
  for (const line of lines.slice(0, 20)) {
    if (
      /^[A-ZÁÉÍÓÚÑ]{2,}(\s+[A-ZÁÉÍÓÚÑ]{2,})+$/.test(line) &&
      !/BBVA|VISA|BANCO|LÍMITES|PESOS|DÓLARES|ARGENTINA|OCASA|RESUMEN/i.test(line)
    ) {
      holderName = line.trim();
      break;
    }
  }

  // Dates: label on one line, DD-Mon-YY on the next
  const cierreActM = text.match(/CIERRE\s+ACTUAL[\s\S]{0,20}?(\d{2}-[A-Za-z]{3}-\d{2,4})/i);
  const vencActM   = text.match(/VENCIMIENTO\s+ACTUAL[\s\S]{0,20}?(\d{2}-[A-Za-z]{3}-\d{2,4})/i);
  const cierreAntM = text.match(/CIERRE\s+ANTERIOR[\s\S]{0,20}?(\d{2}-[A-Za-z]{3}-\d{2,4})/i);

  const today = new Date().toISOString().slice(0, 10);
  return {
    bank_name: "BBVA",
    holder_name: holderName,
    account_number: accountNumber,
    card_last_four: cardLastFour,
    card_network: "Visa",
    period_start: (cierreAntM?.[1] && parseDate(cierreAntM[1])) ?? today,
    period_end:   (cierreActM?.[1]  && parseDate(cierreActM[1]))  ?? today,
    due_date:     (vencActM?.[1]    && parseDate(vencActM[1]))    ?? today,
  };
}

// ─── Balance ──────────────────────────────────────────────────────────────────

function parseBalance(lines: string[]): ParsedBalanceSummary {
  // "DB IVA $ 21%" and "IIBB PERCEP" lines have the base amount on the SAME line
  // and the actual tax on the NEXT line → nextLineOnly=true
  const ivaTax  = sumAmt(lines, /DB\s+IVA\s+\$/i,    true)
                + sumAmt(lines, /IVA\s+RG\s+\d/i,    true);
  const iibbTax = sumAmt(lines, /IIBB\s+PERCEP/i,    true)
                + sumAmt(lines, /DB\.RG/i,            true);
  const selloTax = sumAmt(lines, /IMPUESTO\s+DE\s+SELLOS/i, true);

  // "SALDO ACTUAL U$S\nX,XX" → USD balance
  let currentBalanceUsd: number | undefined;
  for (let i = 0; i < lines.length; i++) {
    if (/^SALDO\s+ACTUAL\s+U\$S$/i.test(lines[i]) && i + 1 < lines.length) {
      const m = lines[i + 1].match(/^(\d{1,3}(?:\.\d{3})*,\d{2})/);
      if (m) { currentBalanceUsd = parseARS(m[1]); break; }
    }
  }

  return {
    currency: "ARS",
    // "SALDO ANTERIOR628.955,942,99" — no space before amount, same-line extraction
    previous_balance:       findAmt(lines, /SALDO\s+ANTERIOR/i),
    // "SU PAGO EN PESOS\n-300.000,00" — negative amount on next line
    payments_applied:       findAmt(lines, /SU\s+PAGO\s+EN\s+PESOS/i,       true),
    // "TOTAL CONSUMOS DE TOMAS GONCALVES\n570.498,282,99" — first AMT on next line
    total_consumption:      findAmt(lines, /TOTAL\s+CONSUMOS/i,              true),
    commission_cuenta_full: findAmt(lines, /COMISI[OÓ]N\s+CUENTA\s+FULL/i,  true),
    sello_tax:              selloTax,
    iva_tax:                ivaTax,
    iibb_tax:               iibbTax,
    financing_interest:     findAmt(lines, /INTERESES\s+FINANCIACION/i,      true),
    // Standalone "SALDO ACTUAL" line (no $ or U$S suffix) → next line is ARS balance
    current_balance:        findAmt(lines, /^SALDO\s+ACTUAL$/i,              true),
    current_balance_usd:    currentBalanceUsd,
    minimum_payment:        findAmt(lines, /PAGO\s+M[IÍ]NIMO/i,             true),
    tna_ars: findPct(lines, /^TNA\s+\$$/i),
    tem_ars: findPct(lines, /^TEM\s+\$$/i),
    tea_ars: findPct(lines, /^TEA\s+\$$/i),
  };
}

// ─── Transactions ─────────────────────────────────────────────────────────────

// BBVA PDF format: date on its own line, description on next, amount on the line after.
const DATE_ONLY   = /^(\d{2}[\/\-](?:\d{2}|[A-Za-z]{3})[\/\-]\d{2,4})$/;
const DATE_INLINE = /^(\d{2}[\/\-](?:\d{2}|[A-Za-z]{3})[\/\-]\d{2,4})\s+(.+)/;
const INSTALLMENT = /\b(?:CTA\s+)?(\d{1,2})\/(\d{1,2})\b/;
const TRAILING_AMT = /(\d{1,3}(?:\.\d{3})*,\d{2})\s*$/;
const TRAILING_NUM = /\s(\d{6,})$/;
const AMT_ONLY    = /^(-?)(\d{1,3}(?:\.\d{3})*,\d{2})(?:[^,\d].*)?$/;

// Lines to skip when they appear as the description of a parsed 3-line block
const DESC_SKIP = /^(FECHA|DESCRIPC|IMPORTE|TOTAL|SU PAGO|TRANSFERENCIA DEUDA|IMPUESTO|INTERESES|COMISION|IVA|IIBB|DB\s|SALDO|Legales|Canales|Sobre\s)/i;

// Section markers
const SECTION_CONSUME_START = /^Consumos\s+\w/i;
const SECTION_CONSUME_END   = /^Impuestos,?\s+cargos/i;

function extractTx(
  date: string,
  descLine: string,
  amtLine: string,
): ParsedTransaction | null {
  const rest0 = descLine.trim();
  if (DESC_SKIP.test(rest0) || rest0.length < 2) return null;

  let amount_ars: number;
  let amount_usd: number | undefined;

  // Try dedicated amount line first
  const amtM = amtLine?.match(AMT_ONLY);
  if (amtM) {
    amount_ars = (amtM[1] === "-" ? -1 : 1) * parseARS(amtM[2]);
    // Handle dual-currency lines like "4.305,60-2,99"
    const usdPart = amtLine.match(/-(\d{1,3}(?:\.\d{3})*,\d{2})\s*$/);
    if (usdPart) amount_usd = parseARS(usdPart[1]);
  } else {
    const trailingM = rest0.match(TRAILING_AMT);
    if (!trailingM) return null;
    amount_ars = parseARS(trailingM[1]);
  }

  // Skip zero-amount lines that are just headers/totals
  if (amount_ars === 0 && !amount_usd) return null;

  let rest = rest0;
  // Remove amount if it was trailing on the description (fallback path)
  if (!amtLine?.match(AMT_ONLY)) {
    const trailingM = rest.match(TRAILING_AMT);
    if (trailingM) rest = rest.slice(0, rest.lastIndexOf(trailingM[1])).trim();
  }

  // Detect USD transactions from description (e.g. "APPLE.COM BILL ... MT575T ... USD 2,99")
  if (/USD|U\$S/i.test(rest) && amount_ars > 0 && amount_ars < 100) {
    amount_usd = amount_ars;
    amount_ars = 0;
  }

  // Installment: "CTA 06/06" or "06/06"
  let installment_current: number | undefined;
  let installment_total: number | undefined;
  const im = rest.match(INSTALLMENT);
  if (im) {
    installment_current = parseInt(im[1], 10);
    installment_total   = parseInt(im[2], 10);
    rest = rest.replace(im[0], "").trim();
  }

  // Trailing voucher number (6+ digits at end, preceded by space)
  let voucher_number: string | undefined;
  const vm = rest.match(TRAILING_NUM);
  if (vm) {
    voucher_number = vm[1];
    rest = rest.slice(0, rest.lastIndexOf(vm[1])).trim();
  }

  const merchant_name = rest.trim();
  if (!merchant_name || merchant_name.length < 2) return null;

  return {
    date,
    merchant_name,
    voucher_number,
    installment_current,
    installment_total,
    amount_ars,
    amount_usd: amount_usd && amount_usd > 0 ? amount_usd : undefined,
  };
}

function parseTransactions(lines: string[]): ParsedTransaction[] {
  const txs: ParsedTransaction[] = [];
  let inConsumos = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Track sections — only include "Consumos" section, skip payments and fees
    if (SECTION_CONSUME_START.test(line)) { inConsumos = true; continue; }
    if (SECTION_CONSUME_END.test(line))   { inConsumos = false; continue; }
    if (!inConsumos) continue;

    // 3-line format: date alone on line i, description on i+1, amount on i+2
    const dm = line.match(DATE_ONLY);
    if (dm) {
      const date = parseDate(dm[1]);
      if (date && i + 2 < lines.length) {
        const tx = extractTx(date, lines[i + 1], lines[i + 2]);
        if (tx) { txs.push(tx); i += 2; continue; }
      }
    }

    // Fallback: inline "date  description  amount" on same line
    const lm = line.match(DATE_INLINE);
    if (lm) {
      const date = parseDate(lm[1]);
      if (!date) continue;
      const tx = extractTx(date, lm[2], "");
      if (tx) txs.push(tx);
    }
  }

  return txs;
}
